validator/validator.go

```diff
@@ -0,0 +1,59 @@
+// Package validator provides functionality for validating groups and units.
+package validator
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/giantswarm/inago/controller"
+)
+
+// ValidateRequest takes a Request, and returns whether it is valid or not.
+// If the request is not valid, the error provides more details.
+func ValidateRequest(request controller.Request) (bool, error) {
+	// Check there are units in the group.
+	if len(request.Units) == 0 {
+		return false, noUnitsInGroupError
+	}
+
+	// Check that each unit name is prefixed with the group name.
+	for _, unit := range request.Units {
+		if !strings.HasPrefix(unit.Name, request.Group) {
+			return false, badUnitPrefixError
+		}
+	}
+
+	// Check that all units either have @ or they don't.
+	numUnitsWithAtSymbol := 0
+	for _, unit := range request.Units {
+		if strings.Contains(unit.Name, "@.") {
```

Comment:
Things like this can be improved by separating the code into helper methods. Something like `unitsContain([]string, string) bool`. That way you can call it `if unitsContain(units, "@") { ... }`. In the helper method you just return true as soon as you see the string within a unit name. On the caller side you simply assert the condition against true or false. No counting necessary.